### PR TITLE
Wallet coins details

### PR DESCRIPTION
**Pull request by Jibz1**
- Implemented coins details page
- moved token page content into a share component, used in the tokens page and token details page

https://user-images.githubusercontent.com/8676844/198157495-6d9e93c7-f339-448d-ad59-8d64e60b3cbf.mov

### CODE REVIEW

File: apps/wallet/src/ui/app/pages/home/tokens/TokensDetails.tsx [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+                    balances={balances}
+                    loading={loading}
+                />
+            ) : (
+                <>
+                    <div className={cl([st.title, st.tokenActivities])}>
+                        {coinSymbol} activity
+                    </div>
+                    <div className={st.txContent}>
+                        <section className={st.activity}>
```
Comment (pchrysochoidis) [inline apps/wallet/src/ui/app/pages/home/tokens/TokensDetails.tsx:159]: why do we need the extra section element?

**Review by mystie711 — CHANGES_REQUESTED** [review]
@Jibz1 please remove the on-hover row highlight blue background

Also, can you show an example with another custom coin in the list and the view for it? I would also like to see the activity view for that coin once in level 2.

Also, making an observation that without the slide in animation from level 1 (home) to 2 (coin detail)...and reverse animation when clicking on the back arrow...the interation feels very awkward. Any chance we could add this animation?

Also, for the tab interaction to work right...anytime when you are deeper in the navigation hierarchy, like being on level 2, tapping on the tab icon should bring the user back to level 1 (home).

**Comment by Jibz1** [comment]
> @Jibz1 please remove the on-hover row highlight blue background
> 
> Can you show an example with another custom coin in the list and the view for it? I would also like to see the activity view for that coin once in level 2.
> 
> Also, making an observation that without the slide in animation from level 1 (home) to 2 (coin detail)...and reverse animation when clicking on the back arrow...the interation feels very awkward. Any chance we could add this animation?
> 
> For the tab interaction to work right...anytime when you are deeper in the navigation hierarchy, like being on level 2, tapping on the tab icon should bring the user back to level 1 (home).



> @Jibz1 please remove the on-hover row highlight blue background
> 
> Can you show an example with another custom coin in the list and the view for it? I would also like to see the activity view for that coin once in level 2.
> 
> Also, making an observation that without the slide in animation from level 1 (home) to 2 (coin detail)...and reverse animation when clicking on the back arrow...the interation feels very awkward. Any chance we could add this animation?
> 
> For the tab interaction to work right...anytime when you are deeper in the navigation hierarchy, like being on level 2, tapping on the tab icon should bring the user back to level 1 (home).

I will try to get some custom tokens.

File: apps/wallet/src/ui/app/pages/home/tokens/TokensPage.module.scss [hunk 2]
```diff
@@ -103,3 +103,22 @@
...
+.activity {
+    margin-top: 10px;
+}
+
+.details-page {
+    overflow-y: hidden;
+}
+
+.tx-content {
+    height: 210px;
```
Comment (pchrysochoidis) [inline apps/wallet/src/ui/app/pages/home/tokens/TokensPage.module.scss:116]: I think we should make this auto expand to the remaining height of the `.container` (I think adding `flex:1` here and the `.container` should do it and use [override-main-padding](https://github.com/MystenLabs/sui/blob/18563ec89fcab0a7b39fc0011be17b759a1a3afb/apps/wallet/src/ui/styles/utils/index.scss#L24) for spacing

File: apps/wallet/src/ui/app/pages/home/tokens/TokensDetails.tsx [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+                    text="Send"
+                />
+                <IconLink
+                    icon={SuiIcons.Swap}
+                    to="/"
+                    disabled={true}
+                    text="Swap"
+                />
+            </div>
+            <div className={st.staking}>
```
Comment (pchrysochoidis) [inline apps/wallet/src/ui/app/pages/home/tokens/TokensDetails.tsx:138]: We should hide this when the coin is not SUI (based on the design)

File: apps/wallet/src/ui/app/pages/home/tokens/TokensDetails.tsx [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+);
+
+type TokensProps = {
+    allCoinTypes: string[];
+    suiBalance: bigint;
+    balances: Record<string, bigint>;
+    loading: boolean;
+};
+
+function MyTokens({
```
Comment (pchrysochoidis) [inline apps/wallet/src/ui/app/pages/home/tokens/TokensDetails.tsx:39]: I would move this to a separate file and probably use store for data - but if we can do it in another PR

File: apps/wallet/src/ui/app/pages/home/tokens/TokenDetailsPage.tsx [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+        [coinType]
+    );
+
+    if (!coinType) {
+        return <Navigate to="/tokens" replace={true} />;
+    }
+    return (
+        <div className={st.detailsPage}>
+            <PageTitle title={symbol} backLink="/tokens" hideBackLabel={true} />
+            <TokenDetails coinType={GAS_TYPE_ARG} />
```
Comment (pchrysochoidis) [inline apps/wallet/src/ui/app/pages/home/tokens/TokenDetailsPage.tsx]: should this be coinType?

File: apps/wallet/src/ui/app/pages/home/tokens/TokensDetails.tsx [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+                <Alert className={st.alert}>
+                    <strong>Sync error (data might be outdated).</strong>{' '}
+                    <small>{error.message}</small>
+                </Alert>
+            ) : null}
+            {!coinType && <AccountAddress showLink={false} mode="faded" />}
+            <div className={st.balanceContainer}>
+                <Loading loading={loading}>
+                    <CoinBalance
+                        balance={suiBalance}
```
[the remark this answers is not in the document]
Comment (Jibz1) [reply apps/wallet/src/ui/app/pages/home/tokens/TokensDetails.tsx:106]: Good catch. Changing the var name

File: apps/wallet/src/ui/app/pages/home/tokens/TokenDetailsPage.tsx (hunk 5 above)
Comment (Jibz1) [reply apps/wallet/src/ui/app/pages/home/tokens/TokenDetailsPage.tsx]: this is was corrected the last commit

File: apps/wallet/src/ui/app/components/transactions-card/TransactionsCard.module.scss [hunk 7]
```diff
@@ -190,3 +190,19 @@
...
 }
+
+.center-loading {
+    display: flex;
+    justify-content: center;
+    align-items: center;
+    height: 100%;
+}
+
+.tx-content {
```
Comment (pchrysochoidis) [inline apps/wallet/src/ui/app/components/transactions-card/TransactionsCard.module.scss:201]: I am not sure we need this, there is this logic/styling already in the parent components

File: apps/wallet/src/ui/app/components/transactions-card/RecentTransactions.tsx [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+    );
+
+    useEffect(() => {
+        dispatch(getTransactionsByAddress()).unwrap();
+    }, [dispatch]);
+
+    return (
+        <>
+            <Loading loading={loading} className={st.centerLoading}>
+                {txByAddress && txByAddress.length ? (
```
Comment (pchrysochoidis) [inline apps/wallet/src/ui/app/components/transactions-card/RecentTransactions.tsx:38]: Not sure is the check is necessary and the section. Would only .map(...) work?

File: apps/wallet/src/ui/app/pages/home/tokens/TokensDetails.tsx [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+                />
+                <IconLink
+                    icon={SuiIcons.Swap}
+                    to="/"
+                    disabled={true}
+                    text="Swap"
+                />
+            </div>
+
+            {coinType === GAS_TYPE_ARG || !coinType ? (
```
Comment (pchrysochoidis) [inline apps/wallet/src/ui/app/pages/home/tokens/TokensDetails.tsx:139]:
```suggestion
            {activeCoinType === GAS_TYPE_ARG ? (
```
should be the same right?

File: apps/wallet/src/ui/app/pages/home/tokens/TokensDetails.tsx [hunk 10]
```diff
@@ -0,0 +1,173 @@
...
+}
+
+function TokenDetails({ coinType }: TokenDetailsProps) {
+    const { loading, error, showError } = useObjectsState();
+    const activeCoinType = coinType || GAS_TYPE_ARG;
+    const balances = useAppSelector(accountAggregateBalancesSelector);
+    const tokenBalance = balances[activeCoinType] || BigInt(0);
+    const allCoinTypes = useMemo(() => Object.keys(balances), [balances]);
+    const coinTypeWithBalance =
+        tokenBalance > 0 ? activeCoinType : allCoinTypes[0];
```
Comment (pchrysochoidis) [inline apps/wallet/src/ui/app/pages/home/tokens/TokensDetails.tsx:88]: I think when we are on the token details page we shouldn't switch to another coin.

Comment (Jibz1) [reply apps/wallet/src/ui/app/pages/home/tokens/TokensDetails.tsx:88]: You are right. I just kept most of the code from the default token page

**Comment by Jibz1** [comment]
@mystie711 please review the video update. I will create a separate PR for the animation.

**Comment by mystie711** [comment]
Thank you for the Slack huddle @Jibz1 .
Let's break the work into two parts if needed to timebox this effort appropriately.

**Comment by mystie711** [comment]
Looking good.

**Review by mystie711 — APPROVED** [review]

File: apps/wallet/src/ui/app/pages/home/tokens/TokensDetails.tsx [hunk 11]
```diff
@@ -0,0 +1,183 @@
...
+}
+
+function TokenDetails({ coinType }: TokenDetailsProps) {
+    const { loading, error, showError } = useObjectsState();
+    const activeCoinType = coinType || GAS_TYPE_ARG;
+    const balances = useAppSelector(accountAggregateBalancesSelector);
+    const tokenBalance = balances[activeCoinType] || BigInt(0);
+    const allCoinTypes = useMemo(() => Object.keys(balances), [balances]);
+    const coinTypeWithBalance =
+        !coinType && tokenBalance > 0 ? activeCoinType : allCoinTypes[0];
```
Comment (pchrysochoidis) [inline apps/wallet/src/ui/app/pages/home/tokens/TokensDetails.tsx:89]:
```suggestion
        coinType || tokenBalance > 0 ? activeCoinType : allCoinTypes[0];
```

Otherwise I think when we provide a coinType there is a chance to select a different coin

File: apps/wallet/src/ui/app/pages/home/tokens/TokensDetails.tsx [hunk 12]
```diff
@@ -0,0 +1,183 @@
...
+                        to="/"
+                        disabled={true}
+                        text="Buy"
+                    />
+                    <IconLink
+                        icon={SuiIcons.ArrowLeft}
+                        to={`/send${
+                            coinTypeWithBalance
+                                ? `?${new URLSearchParams({
+                                      type: activeCoinType,
```
Comment (pchrysochoidis) [inline apps/wallet/src/ui/app/pages/home/tokens/TokensDetails.tsx:135]:
```suggestion
                                      type: coinTypeWithBalance,
```

**Review by pchrysochoidis — APPROVED** [review]
